sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -56,33 +58,130 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
-          val eval = e.genCode(ctx)
-          eval.code + s"""
-            if (${eval.isNull}) {
-              $values[$i] = null;
-            } else {
-              $values[$i] = ${eval.value};
-            }
-           """
-        }) +
-      s"""
-        final ArrayData ${ev.value} = new $arrayClass($values);
-        this.$values = null;
-      """, isNull = "false")
+    val et = dataType.elementType
+    val evals = children.map(e => e.genCode(ctx))
+    val isPrimitiveArray = ctx.isPrimitiveType(et)
+    val (preprocess, postprocess, arrayData, array) =
+      GenArrayData.genCodeToCreateArrayData(ctx, et, children.size, isPrimitiveArray)
+    val assigns = GenArrayData.genCodeToAssignArrayElements(
+      ctx, evals, et, isPrimitiveArray, arrayData, array, true)
+    /*
+      TODO: When we generate simpler code, we have to solve the following exception
+        https://github.com/apache/spark/pull/13909/files#r93813725
+      ev.copy(
+        code = preprocess + ctx.splitExpressions(ctx.INPUT_ROW, assigns) + postprocess
+        value = arrayData,
+        isNull = "false")
+    */
+    ev.copy(
+      code = preprocess + ctx.splitExpressions(ctx.INPUT_ROW, assigns) + postprocess +
+        s"\nfinal ArrayData ${ev.value} = $arrayData;\n",
+      isNull = "false")
   }
 
   override def prettyName: String = "array"
 }
 
+private [sql] object GenArrayData {
+  /**
+   * Return Java code pieces based on DataType and isPrimitive to allocate ArrayData class
+   *
+   * @param ctx a [[CodegenContext]]
+   * @param elementType data type of an underlying array
+   * @param numElements the number of array elements
+   * @param isPrimitive Are all of the elements of an underlying array primitive type
+   * @return (code pre-assignments, code post-assignments, underlying array name, arrayData name)
+   */
+  def genCodeToCreateArrayData(
+      ctx: CodegenContext,
+      elementType: DataType,
+      numElements: Int,
+      isPrimitive : Boolean): (String, String, String, String) = {
+    val arrayName = ctx.freshName("array")
+    val arrayDataName = ctx.freshName("arrayData")
+    if (!isPrimitive) {
+      val arrayClass = classOf[GenericArrayData].getName
+      ctx.addMutableState("Object[]", arrayName,
+        s"this.$arrayName = new Object[${numElements}];")
+      ("",
+       s"$arrayClass $arrayDataName = new $arrayClass($arrayName);",
+       arrayDataName,
+       arrayName)
+    } else {
+      val unsafeArrayClass = classOf[UnsafeArrayData].getName
+      val baseObject = ctx.freshName("baseObject")
+      val unsafeArraySizeInBytes =
+        UnsafeArrayData.calculateHeaderPortionInBytes(numElements) +
+        ByteArrayMethods.roundNumberOfBytesToNearestWord(elementType.defaultSize * numElements)
+      val baseOffset = Platform.BYTE_ARRAY_OFFSET
+
+      (s"""
+        byte[] $arrayName = new byte[$unsafeArraySizeInBytes];
+        $unsafeArrayClass $arrayDataName = new $unsafeArrayClass();
```

Comment:
I think @ueshin meant the split element assignment codes won't access this local `arrayDataName`. So you might need to use `addMutableState` to add them in.